[ENH] change `_MetaObjectMixin._check_objects` default for `attr_name` to auto-detect

SimonBlanke opened the pull request:
#### Reference Issues/PRs

Closes #159


#### What does this implement/fix? Explain your changes.
<!--
A clear and concise description of what you have implemented. Remember to implement
unit tests and docstrings if your pull request commits code to the repository.
-->

Sets `attr_name` to `None` and auto detect from "named_object_parameters" of `attr_name` remains `None`

#### Does your contribution introduce a new dependency? If yes, which one?

<!--
If your contribution requires a new dependency please indicate why it is necessary.
skbase seeks to minimize dependencies to make it easy to use skbase in a variety
of environments and contexts.
-->

#### What should a reviewer concentrate their feedback on?

<!-- This section is particularly useful if you have a pull request that is still in development.
You can guide the reviews to focus on the parts that are ready for their comments.
We suggest using bullets (indicated by * or -) and filled checkboxes [x] here -->

#### Any other comments?
<!--
Is there any other information the reviewer should know?
-->

#### PR checklist
<!--
Please go through the checklist below. Please feel free to remove points if they are not applicable.
-->

##### For all contributions
- [ ] I've reviewed the project documentation on [contributing](https://skbase.readthedocs.io/en/latest/contribute.html)
- [ ] I've added myself to the [list of contributors](https://github.com/sktime/skbase/blob/main/.all-contributorsrc).
- [ ] The PR title starts with either [ENH], [CI/CD], [MNT], [DOC], or [BUG] indicating whether
  the PR topic is related to enhancement, CI/CD, maintenance, documentation, or a bug.

##### For code contributions
- [ ] Unit tests have been added covering code functionality
- [ ] Appropriate docstrings have been added (see [documentation standards](https://skbase.readthedocs.io/en/latest/contribute/development/developer_guide/creating_docs.html))
- [ ] New public functionality has been added to the API Reference


<!--
Thanks for contributing!
-->

SimonBlanke committed:
change default of attr_name

SimonBlanke committed:
Auto-detect attr_name from the named_object_parameters tag if not provided

SimonBlanke committed:
add 'test_check_objects_attr_name_auto_detection'

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

SimonBlanke committed:
add test 'test_check_objects_attr_name_explicit'

SimonBlanke committed:
merge

SimonBlanke committed:
add test 'test_check_objects_attr_name_custom_tag'

SimonBlanke commented:
added another test 'test_check_objects_attr_name_custom_tag'. It sets a custom tag via `set_tags`.

I just realized, that I used a private method `_check_objects` for the tests. Normally, that should be a no-go, but it seems the method is not used anywhere else in the package.

SimonBlanke committed:
improve test asserts by triggering error

SimonBlanke committed:
Merge branch 'main' into feature/change-default

SimonBlanke committed:
merge main into pr

fkiraly reviewed (changes requested):
I think the coupling to `self` should happen at least one layer higher, i.e., not in the checker function, but where it is being called.

Otherwise we have intransparent coupling in a function that looks like a pure function but actually is not.

SimonBlanke commented:
> I think the coupling to `self` should happen at least one layer higher, i.e., not in the checker function, but where it is being called.
> 
> Otherwise we have intransparent coupling in a function that looks like a pure function but actually is not.

I understand we have some hidden coupling going on in `_check_objects`. The method signature does not reveal how it works. The tags are surprisingly read without the user knowing. From your explanation I infer the problem is, that some high level orchestration stuff (object stats, method calls) is mixed with lower level logic. It would help if we separate this.

I'll think about how to solve this.

SimonBlanke commented:
> are there any other tags that get read in there?

No, I only see `attr_name = self.get_tag("named_object_parameters")` here. For the decoupling I would like to separate the code into the following methods: 

```python
def _check_objects(self, objs, attr_name=None, ...):
    """Orchestration/high-level layer."""
    if attr_name is None:
        attr_name = self.get_tag("named_object_parameters")

    validated = self._validate_objects(objs, attr_name, ...)
    return self._coerce_to_named_object_tuples(validated, ...)

@staticmethod
def _validate_objects(objs, attr_name, cls_type=None, ...):
    # All validation logic
```

SimonBlanke committed:
Merge branch 'main' into feature/change-default

SimonBlanke committed:
make '_validate_objects' staticmethod

SimonBlanke committed:
add some comments... because this is getting complex

SimonBlanke committed:
add docstring

SimonBlanke committed:
add docstring

SimonBlanke committed:
Merge branch 'main' into feature/change-default

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

SimonBlanke commented:
okay there is a lot going on here (or maybe this is just in my head). First I added [this note](https://github.com/sktime/skbase/pull/466/files#diff-e6197ccf8ea00f7080233f04b26c12e30a6e790958541927d724179387010da5R390), because `_check_objects` does not really make sense to me as a private method, right now.

`_validate_objects` is a staticmethod now containing all the validation logic, `_check_objects` handles the rest. I also added a lot of comments. It feels necessary here.

SimonBlanke committed:
add 'test_check_objects_attr_name_none_tag_not_set'

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

SimonBlanke committed:
fix linter error

SimonBlanke committed:
Merge branch 'feature/change-default' of https://github.com/SimonBlanke/skbase into feature/change-default

SimonBlanke commented:
> There is a linter error in a file I did not change: "skbase/tests/test_base.py:389:5: B043 Do not call delattr with a constant attribute value, it is not any safer than normal property access."

fixed!

SimonBlanke committed:
Merge branch 'main' into feature/change-default

fkiraly committed:
Update _dependencies.py

fkiraly committed:
Merge branch 'main' into pr/466

fkiraly reviewed (approved):
Looks good to me!
